src/java.base/share/classes/sun/security/provider/NamedSignature.java

```diff
@@ -0,0 +1,222 @@
+    private byte[] pubKey = null;
+
+    private Object sk2 = null;
+    private Object pk2 = null;
+
+    /// Creates a new `NamedSignature` object.
+    ///
+    /// @param fname the family name
+    /// @param pnames the standard parameter set names, at least one is needed.
+    protected NamedSignature(String fname, String... pnames) {
+        if (fname == null) {
+            throw new AssertionError("fname cannot be null");
+        }
+        if (pnames == null || pnames.length == 0) {
+            throw new AssertionError("pnames cannot be null or empty");
+        }
+        this.fname = fname;
+        this.pnames = pnames;
+    }
+
+    @Override
+    protected void engineInitVerify(PublicKey publicKey) throws InvalidKeyException {
+        // translate also check the key
+        var nk = (NamedX509Key) new NamedKeyFactory(fname, pnames)
+                .engineTranslateKey(publicKey);
+        name = nk.getParams().getName();
+        pubKey = nk.getRawBytes();
+        pk2 = implCheckPublicKey(name, pubKey);
+        secKey = null;
+        bout.reset();
+    }
+
+    @Override
+    protected void engineInitSign(PrivateKey privateKey) throws InvalidKeyException {
+        // translate also check the key
+        var nk = (NamedPKCS8Key) new NamedKeyFactory(fname, pnames)
+                .engineTranslateKey(privateKey);
+        name = nk.getParams().getName();
+        secKey = nk.getRawBytes();
+        sk2 = implCheckPrivateKey(name, secKey);
+        pubKey = null;
+        bout.reset();
+    }
+
+    @Override
+    protected void engineUpdate(byte b) throws SignatureException {
+        bout.write(b);
+    }
+
+    @Override
+    protected void engineUpdate(byte[] b, int off, int len) throws SignatureException {
+        bout.write(b, off, len);
+    }
+
+    @Override
+    protected byte[] engineSign() throws SignatureException {
+        if (secKey != null) {
+            var msg = bout.toByteArray();
+            bout.reset();
+            return implSign(name, secKey, sk2, msg, appRandom);
+        } else {
+            throw new IllegalStateException("No private key");
+        }
+    }
+
+    @Override
+    protected boolean engineVerify(byte[] sig) throws SignatureException {
+        if (pubKey != null) {
+            var msg = bout.toByteArray();
+            bout.reset();
+            return implVerify(name, pubKey, pk2, msg, sig);
+        } else {
+            throw new IllegalStateException("No public key");
+        }
+    }
+
+    @Override
+    @SuppressWarnings("deprecation")
+    protected void engineSetParameter(String param, Object value)
+            throws InvalidParameterException {
+        throw new UnsupportedOperationException("setParameter() not supported");
+    }
+
+    @Override
+    @SuppressWarnings("deprecation")
+    protected Object engineGetParameter(String param) throws InvalidParameterException {
+        throw new UnsupportedOperationException("getParameter() not supported");
```

Comment:
`engineGetParameter` is not specified to throw UOE, so suggest throwing `InvalidParameterException` instead. Same comment for `engineSetParameter`.

Reply:
Yes I can. I noticed that the newest implementation `HSS` does throw `InvalidParameterException`, and a little older one `ECDSASignature` and `EdDSASignature` were throwing `UnsupportedOperationException`.